### ISSUE TITLE
Question about Target class sampling

### ISSUE THREAD

**Issue by JiaDingCN** [bhheo/BSS_distillation#3]
Hi, I'm reading your code and a little confused about the Target class sampling part. In paper you say the choice of class k is based on their probabilities of the teacher, that is to say, a more discrinate class gets a higher probability to be chosen. But in your code you use:
```Python
class_score, class_idx = F.softmax(output_te, 1)[attack_idx, :].data.sort(dim=1, descending=True)
class_score = class_score[:, 1:]
class_idx = class_idx[:, 1:]
rand_seed = 1 * (class_score.sum(dim=1) * torch.rand([attack_idx.shape[0]]).cuda()).unsqueeze(1)
prob = class_score.cumsum(dim=1)
for k in range(attack_idx.shape[0]):
    for c in range(prob.shape[1]):
    if (prob[k, c] >= rand_seed[k]).cpu().numpy():
        attack_class[k] = class_idx[k, c]
        break
```
So for  a sample `k`, you just set a bar ` rand_seed[k])` and sort the probabilities in `prob` with the minimum probility one in the beginning. Then you select from the beginning in `prob` and exam whether it exceeds the bar. I wonder if my understanding is correct. If so, it may cause problem since the classes with the highest probilities are listed in the end of the `prob` and they are less likely to be selected because the selection begins in the begining, which does not match what the paper claims.

I wonder why you do not just divide [0,1] to different classes and make a random number between 0 and 1. It seems to be easier and correct.

**Comment by bhheo** [comment]
Hi

Please consider that the `rand_seed` is multiplication of random in [0, 1] and `class_score.sum(dim=1)`.
So, the maximum value of `rand_seed` is the same as the last value of `prob`.
All classes could be chosen with the probability proportional to their `prob`.
It is a typical implementation of probabilistic sampling using the cumulative sum.
The problem you mentioned won't happen. It is the correct implementation.

As you said, I can just use 
```python
rand_seed = torch.rand([attack_idx.shape[0]]).cuda().unsqueeze(1)
prob = class_score.cumsum(dim=1) / class_score.sum(dim=1)
```
It works the same as the current implementation except for special cases.
However,  the class scores except for maximum prob `class_score = class_score[:, 1:]` usually have very small values on classification training.
A division with the small value `1 / class_score.sum(dim=1)` has a risk to makes overflow errors such as divide by zero or `NaN`.
So, I choose an alternative implementation without the division.

**Comment by JiaDingCN** [comment]
Oh, I get it. I misunderstood the meaning of `prob` before. It makes sense if `prob` is the cumulative sum.

Sorry for taking up your time. Thank you for your interesting work and warm help!
